packages/piro/test/MockModelEval_C_Tpetra.cpp

```diff
@@ -0,0 +1,535 @@
+
+
+  const Teuchos::RCP<const Tpetra_Vector> lag_multiplier_g_in =
+      Teuchos::nonnull(inArgs.get_g_multiplier(0)) ?
+        ConverterT::getConstTpetraVector(inArgs.get_g_multiplier(0)) :
+        Teuchos::null;
+
+  auto g_hess_xx_v = outArgs.supports(Thyra::ModelEvaluator<double>::OUT_ARG_hess_vec_prod_g_xx,0) ? outArgs.get_hess_vec_prod_g_xx(0) : Teuchos::null;
+  const Teuchos::RCP<Tpetra_MultiVector> g_hess_xx_v_out =
+      Teuchos::nonnull(g_hess_xx_v) ?
+        ConverterT::getTpetraMultiVector(g_hess_xx_v) :
+        Teuchos::null;
+
+  auto g_hess_xp_v = outArgs.supports(Thyra::ModelEvaluator<double>::OUT_ARG_hess_vec_prod_g_xp,0,0) ? outArgs.get_hess_vec_prod_g_xp(0,0) : Teuchos::null;
+  const Teuchos::RCP<Tpetra_MultiVector> g_hess_xp_v_out =
+      Teuchos::nonnull(g_hess_xp_v) ?
+        ConverterT::getTpetraMultiVector(g_hess_xp_v) :
+        Teuchos::null;
+
+  auto g_hess_px_v = outArgs.supports(Thyra::ModelEvaluator<double>::OUT_ARG_hess_vec_prod_g_px,0,0) ? outArgs.get_hess_vec_prod_g_px(0,0) : Teuchos::null;
+  const Teuchos::RCP<Tpetra_MultiVector> g_hess_px_v_out =
+      Teuchos::nonnull(g_hess_px_v) ?
+        ConverterT::getTpetraMultiVector(g_hess_px_v) :
+        Teuchos::null;
+
+  auto g_hess_pp_v = outArgs.supports(Thyra::ModelEvaluator<double>::OUT_ARG_hess_vec_prod_g_pp,0,0,0) ? outArgs.get_hess_vec_prod_g_pp(0,0,0) : Teuchos::null;
+  const Teuchos::RCP<Tpetra_MultiVector> g_hess_pp_v_out =
+      Teuchos::nonnull(g_hess_pp_v) ?
+        ConverterT::getTpetraMultiVector(g_hess_pp_v) :
+        Teuchos::null;
+
+  Teuchos::RCP<const Tpetra_Vector> x_dotdot_in;
+  double omega(0.0);
+  // This assumes we are using Piro second order integrator 
+  if (Teuchos::nonnull(this->get_x_dotdot())) {
+    const Teuchos::RCP<const Thyra::VectorBase<double>> x_dotdot = this->get_x_dotdot();
+    x_dotdot_in = ConverterT::getConstTpetraVector(x_dotdot);
+    omega    = this->get_omega();
+  }
+
+  // If using Tempus, we should have 
+  //  x_dotdot_in = Tuchos::nonnull(inArgs.get_x_dot_dot()) ?
+  //        ConverterT::getConstTpetraVector(inArgs.get_x_dot_dot()) :
+  //        Teuchos::null;
+  //omega = inArgs.get_W_x_dot_dot_coeff();
+
+  const double damping = 0;
+  if (Teuchos::nonnull(f_out)) {
+    f_out->putScalar(0.0);
+    auto p = p_vec->getData();
+    auto f_out_data = f_out->getDataNonConst();
+    for (int i=0; i<myVecLength; i++)
+      f_out_data[i] = -p[0];
+
+    if (Teuchos::nonnull(x_dotdot_in)) {
+      // f(x, x_dot) = x_dotdot - f(x)
+      auto f_out_data = f_out->getDataNonConst();
+      for (int i=0; i<myVecLength; i++) {
+        f_out_data[i] = x_dotdot_in->getData()[i] - f_out->getData()[i];
+      }
+    }
+
+    if (Teuchos::nonnull(f_out) && Teuchos::nonnull(x_dot_in)) {
+      // f(x, x_dot) = f(x) -  damping * x_dot
+      auto f_out_data = f_out->getDataNonConst();
+      for (int i=0; i<myVecLength; i++) {
+        f_out_data[i] += damping * x_dot_in->getData()[i];
```

Comment:
This and some of the functions below should get the view, not getData() and execute this operation on device. I realize this is a small test problem so the runtime is not an issue, but we should probably show users what they should do for real problems of a very large size. Solution vector operations should be run on device.